pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,306 @@
+	}
+	return true
+}
+
+func (c *MCPodRouteController) enqueuePod(obj interface{}, isDelete bool) {
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		klog.ErrorS(errors.New("received unexpected object"), "enqueuePod can't process event", "obj", obj)
+		return
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	if c.activeGWName == "" || c.nodeConfig.Name != c.activeGWName {
+		klog.InfoS("No active gateway or the Node is not the active Gateway", "node", c.nodeConfig.Name, "gateway", c.activeGWName)
+		return
+	}
+
+	podIP := pod.Status.PodIP
+	podNodeIP := pod.Status.HostIP
+	_, exists := c.installedPods[podIP]
+	if isDelete && exists {
+		c.queue.Add(&endpointChangeInfo{
+			podIP: podIP,
+		})
+		delete(c.installedPods, podIP)
+		return
+	}
+
+	isValidPod := c.validatePod(pod)
+	if !isValidPod {
+		return
+	}
+	c.queue.Add(&endpointChangeInfo{
+		podNodeIP: podNodeIP,
+		podIP:     podIP,
+	})
+	c.installedPods[podIP] = podNodeIP
+}
+
+func (c *MCPodRouteController) Run(stopCh <-chan struct{}) {
+	controllerName := "AntreaAgentMulticlusterNodeRouteController"
+	defer c.queue.ShutDown()
+
+	cacheSyncs := []cache.InformerSynced{c.podInformer.HasSynced, c.gwInformer.HasSynced}
+	klog.InfoS("Starting controller", "controller", controllerName)
+	defer klog.InfoS("Shutting down controller", "controller", controllerName)
+	if !cache.WaitForNamedCacheSync(controllerName, stopCh, cacheSyncs...) {
+		return
+	}
+
+	c.initialize()
+	for i := 0; i < defaultWorkerNum; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (c *MCPodRouteController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *MCPodRouteController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+	if k, ok := obj.(*endpointChangeInfo); !ok {
+		c.queue.Forget(obj)
+		klog.InfoS("Expected endpointChangeInfo in work queue but got", "object", obj)
+		return true
+	} else if err := c.syncPodFlows(k); err == nil {
+		c.queue.Forget(k)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(k)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+	}
+	return true
+}
+
+func (c *MCPodRouteController) initialize() {
+	activeGW, err := getActiveGateway(c.gwLister, c.namespace)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get an active Gateway")
+	}
+
+	if activeGW != nil {
+		c.mutex.Lock()
+		defer c.mutex.Unlock()
+		c.activeGWName = activeGW.Name
+		klog.InfoS("Found an active Gateway", "gateway", klog.KObj(activeGW))
+	}
+}
+
+func (c *MCPodRouteController) syncPodFlows(key *endpointChangeInfo) error {
+	isDelete := key.podNodeIP == ""
+	if isDelete {
+		klog.V(2).InfoS("Deleting Multi-cluster flows for Pod", "podIP", key.podIP)
+		if err := c.ofClient.UninstallMulticlusterPodFlows(key.podIP); err != nil {
+			klog.ErrorS(err, "Failed to uninstall Multi-cluster flows for Pod", "podIP", key.podIP)
+			return err
```

Comment:
If the Pod is deleted, the IP will be assigned to another Pod, then it will have a new Pod creation event to trigger flow sync process. For deletion case, we use Pod IP as the key, so we actually don't need Node IP to delete flow.

Reply:
Let me try explaining the event sequence I meant:

1. Pod deletion tasks is queued.
2. Pod IP is reallocated to another Pod, and a Pod creation talks is queued.
3. UninstallMulticlusterPodFlows failed, and the Pod deletion task is requeued.
4. Pod creation talk is handled.
5. Pod deletion is handled.

Again, let us always use the latest state in cache or K8s. Do not append state to tasks which may become invalid.

Reply:
Done, refined the whole Pod flow controller.